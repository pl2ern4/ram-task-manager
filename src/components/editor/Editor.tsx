import dynamic from 'next/dynamic';
import { styled } from '@mui/material/styles';
import { SetStateAction, useState } from 'react'
import 'react-quill/dist/quill.snow.css';


interface Props {
    label?: string;
    shouldDisplayBorderOnReadOnly: boolean;
    shouldShowButtons?:boolean;
    shouldAlwaysEditable?:boolean;
    setDescription:any;
    value?: string;
}

interface EditorReadOnlyProps {
    shouldDisplayBorderOnReadOnly: boolean
}

const EditorReadOnly = styled('div')<EditorReadOnlyProps>(({shouldDisplayBorderOnReadOnly})=>({
    border: shouldDisplayBorderOnReadOnly? `.5px solid grey`:'',
    minHeight: '3rem',
    width: '100%',
    borderRadius:'.2rem',
    overflow: 'scroll'
}))

const ReactQuill = dynamic(
    () => import('react-quill'),
    { ssr: false }
  )

const Editor = ({shouldDisplayBorderOnReadOnly, label, shouldAlwaysEditable, setDescription, shouldShowButtons, value=''}:Props) => {
    const [editorHtml, setEditorHTML] = useState(value);
    const [content, setContent] = useState(value);
    const [isEditable, setIsEditable] = useState(false);
    const handleChange = (html: SetStateAction<string>) => {
        setEditorHTML(html);
        !shouldShowButtons && setDescription && setDescription(html);
    }
    const handleSubmit = () => {
        setIsEditable(false);
        setContent('');
        setDescription && setDescription(editorHtml);
        setEditorHTML('')
    }
    const handleDoubleClick = () => {
        setIsEditable(true);
    }
    const handleCancel = () =>{
        setIsEditable(false);
    }
    return (
        <div>
            {label && <label><h3>{label} :</h3></label>}
            {!isEditable && !shouldAlwaysEditable? <EditorReadOnly shouldDisplayBorderOnReadOnly={shouldDisplayBorderOnReadOnly} className='clickable-div' onDoubleClick={handleDoubleClick} dangerouslySetInnerHTML={{__html: content }}/> :
                <div>
                        <ReactQuill 
                            onChange={handleChange}
                            value={editorHtml}
                            modules={Editor.modules}
                            formats={Editor.formats}
                            bounds={'.app'}
                        />
                        {
                            shouldShowButtons && (
                                <>
                                    <button type='button' onClick={handleSubmit}>Save</button>   
                                    <button type='button' onClick={handleCancel}>Cancel</button>  
                                </>
                            )
                        }
                </div>
            }
        </div>
    )


}

Editor.modules = {
    toolbar: [
      [{ 'header': '1'}, {'header': '2'}, { 'font': [] }],
      [{size: [200]}],
      ['bold', 'italic', 'underline', 'strike', 'blockquote'],
      [{'list': 'ordered'}, {'list': 'bullet'}, 
       {'indent': '-1'}, {'indent': '+1'}],
       [{ 'color': [] }, { 'background': [] }],
      ['image', 'video'],
      ['clean']
    ]
  }
  /* 
   * Quill editor formats
   * See https://quilljs.com/docs/formats/
   */
  Editor.formats = [
    'header', 'font', 'size',
    'bold', 'italic', 'underline', 'strike', 'blockquote', 'color','background',
    'list', 'bullet', 'indent',
    'image', 'video'
  ]

export default Editor;