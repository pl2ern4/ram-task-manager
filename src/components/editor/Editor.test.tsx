import { render } from '@testing-library/react';
import Editor from './Editor';

describe('Editor', () => {
    const mockFn = jest.fn();
    it('should Match Snapshot', async () => {
        const { container } = render(<Editor setDescription={mockFn} label="Test" value="test value" shouldDisplayBorderOnReadOnly={true} />)
        expect(container).toMatchSnapshot();
    });  
    it('should Match Snapshot, when `shouldDisplayBorderOnReadOnly` passed false', async () => {
        const { container } = render(<Editor setDescription={mockFn} label="Test" value="" shouldDisplayBorderOnReadOnly={false}/>)
        expect(container).toMatchSnapshot();
    }); 
    it('should Match Snapshot, when `label` passed blank', async () => {
        const { container } = render(<Editor setDescription={mockFn} label="" value="" shouldDisplayBorderOnReadOnly={false}/>)
        expect(container).toMatchSnapshot();
    });  
    it('should Match Snapshot, when `shouldAlwaysEditable` passed true', async () => {
        const { container } = render(<Editor shouldDisplayBorderOnReadOnly={true} setDescription={mockFn} label="" value="" shouldAlwaysEditable={true}/>)
        expect(container).toMatchSnapshot();
    });  
})