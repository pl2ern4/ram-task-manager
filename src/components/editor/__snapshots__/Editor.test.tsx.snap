// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should Match Snapshot 1`] = `
<div>
  <div>
    <label>
      <h3>
        Test
         :
      </h3>
    </label>
    <div
      class="clickable-div css-1v42c4z"
    >
      test value
    </div>
  </div>
</div>
`;

exports[`Editor should Match Snapshot, when \`label\` passed blank 1`] = `
<div>
  <div>
    <div
      class="clickable-div css-gs0ilt"
    />
  </div>
</div>
`;

exports[`Editor should Match Snapshot, when \`shouldAlwaysEditable\` passed true 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;

exports[`Editor should Match Snapshot, when \`shouldDisplayBorderOnReadOnly\` passed false 1`] = `
<div>
  <div>
    <label>
      <h3>
        Test
         :
      </h3>
    </label>
    <div
      class="clickable-div css-gs0ilt"
    />
  </div>
</div>
`;
